Add method to hs.eventtap.event to capture touch data for events of type NSEventTypeGesture (29)

asmagill opened the pull request:
I've tested this with my built in touchpad and an external ForceTouch touchpad and it seems to be working well.

Addresses #2470

@CantGetRight82, @latenitefilms:

Some of the documentation in the Apple header files suggest that these events might also be generated by using the touchbar, but my laptop is too old -- I don't have one. I tried with the virtual touchbar module (https://github.com/asmagill/hs._asm.undocumented.touchbar) and swear that I did get a response once that threw an exception saying that "normalizedPosition wasn't a valid method for this type of NSTouch", which is why the `@try/@catch` is in the `NSTouch_toLua` function... but I can't replicate it now.

@latenitefilms I know you have (or at least have access to) a touchbar enabled Mac, so please try this out and let me know what, if anything you observe.

As a a very simple test, I've been using this:

~~~lua
j = hs.eventtap.new({29}, function(e) print(hs.inspect(e:getTouches(), { newline = " ", indent = "" })) end):start()
~~~

I'm most interested in knowing if it works for you, but in addition:

1. does it detect gestures/movements on the touchbar?
2.  and if so, do the `resting` and `type` fields change? (I've only seen them as `false` and `indirect` respectively), and
3. does `normalizedPosition` appear in the tables?

latenitefilms commented:
Amazing! Will test out tonight.

Also, thanks to this [Tweet](https://twitter.com/c8whizper/status/1308049611697917952), I came across [SensibleSideButtons](https://github.com/archagon/sensible-side-buttons), which looks like it can simulate gestures, which may be helpful for #1434.

latenitefilms commented:
Here's my results when trying a bunch of things on my MacBook Pro (15-inch, 2017) Track Pad:

https://gist.github.com/latenitefilms/cf42e16ff2846f7c79c5640251a41c8b

Tapping and dragging on the Touch Bar does generate events, but they're just `nil`.

CantGetRight82 commented:
@asmagill I just tried this and it works like a charm! The code looks neat and you have definitely sparked my enthusiasm in contributing when I have another unimplemented idea.

As for the Touch Bar, I unfortunately do not have it. I looked into try/catch. Try is zero cost but throwing might incur a penalty: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Exceptions/Articles/Exceptions64Bit.html#//apple_ref/doc/uid/TP40009044-SW1
So if I was to have a Touch Bar, I might prefer respondsToSelector instead. Then again... how costly could it really be?

Cant't wait for this branch to be merged to master. Amazing work, thank you!

asmagill commented:
I want to poke at it a little more before wrapping it into the core. As to checking for the selector, usually I would do that, but I only got the crash a few times early on and the exception reason (I wish I had copy/pasted it somewhere, but I didn't) said something along the lines of `normalizedPosition not valid for this type of NSTouch`, so the object *was* an NSTouch (or subclass) and I suspect a check for the selector would have returned YES. If I can recreate the exact conditions in which I encountered it, I will definitely try, though, because you're right -- @try/@catch/@finally is expensive in comparison.

I'd also like to see if I can figure out whether or not it should work with the touchbar (virtual or otherwise)... some of the comments in the header files for `NSTouch.h` suggest that it *should*, but @latenitefilms comments aren't encouraging so far...

latenitefilms commented:
Well, touching the Touch Bar is definitely being detected, as I am seeing the callback being triggered, it's just returning `nil`. I'm happy to do more testing if you want to add in some further debugging code, etc?

latenitefilms commented:
Maybe relevant?

https://www.mail-archive.com/search?l=cocoa-dev@lists.apple.com&q=subject:%22Quartz+Event+Taps+and+Touch+Bar%22&o=newest&f=1

asmagill commented:
Try:

~~~lua
j = hs.eventtap.new({29, 37}, function(e)
    print(hs.inspect(e:getTouches(), { newline = " ", indent = "" }))
 end):start()
~~~

Looks like there are a few new events to add to the `types` list as well.

latenitefilms commented:
I'm still just seeing a whole lot of `nil` when using this branch:

```lua
> j = hs.eventtap.new({29, 37}, function(e)
    print(hs.inspect(e:getTouches(), { newline = " ", indent = "" }))
 end):start()

2020-09-24 09:38:43: nil
2020-09-24 09:38:43: nil
2020-09-24 09:38:44: nil
2020-09-24 09:38:44: nil
2020-09-24 09:38:44: nil
2020-09-24 09:38:44: nil
2020-09-24 09:38:44: nil
```

asmagill commented:
Doh!

https://github.com/Hammerspoon/hammerspoon/pull/2493/files#diff-48caf9bea3f4abbc0620b9457b8583d8R1014 
would also have to be changed to

~~~objc
if (asNSEvent.type == NSEventTypeGesture || asNSEvent.type == NSEventTypeDirectTouch) {
~~~

(this prevents a different type of crash)

asmagill commented:
But I'll look closer at the link you sent earlier as well (that's where I found out about NSEventTypeDirectTouch)

latenitefilms commented:
Victory!

```lua
> j = hs.eventtap.new({29, 37}, function(e)
    print(hs.inspect(e:getTouches(), { newline = " ", indent = "" }))
 end):start()

> -- Single Tap

2020-09-24 10:02:05: -- Loading extension: inspect
2020-09-24 10:02:05: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040004cc4d0", phase = "began", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:05: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040004cc4d0", phase = "ended", resting = false, touching = false, type = "direct" } }

> -- Double Tap

2020-09-24 10:02:10: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000635e90", phase = "began", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:10: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000635e90", phase = "ended", resting = false, touching = false, type = "direct" } }
2020-09-24 10:02:10: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006386d0", phase = "began", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:10: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006386d0", phase = "ended", resting = false, touching = false, type = "direct" } }

> -- Tap and Drag Left

2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "began", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:17: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000690c50", phase = "ended", resting = false, touching = false, type = "direct" } }

> -- Tap and Drag Right

2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "began", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "moved", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:22: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040006f5c10", phase = "ended", resting = false, touching = false, type = "direct" } }

> -- Two Finger Tap

2020-09-24 10:02:43: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040002a22d0", phase = "began", resting = false, touching = true, type = "direct" }, { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000291690", phase = "began", resting = false, touching = true, type = "direct" } }
2020-09-24 10:02:43: { { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x6040002a22d0", phase = "ended", resting = false, touching = false, type = "direct" }, { device = "0x608000166ca0", deviceSize = { __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }, identity = "0x604000291690", phase = "ended", resting = false, touching = false, type = "direct" } }
```

latenitefilms commented:
Some other potential interesting links:

- [Accessing raw multitouch trackpad data (MacOS)](https://gist.github.com/rmhsilva/61cc45587ed34707da34818a76476e11)
- [OpenMultitouchSupport](https://github.com/Kyome22/OpenMultitouchSupport)

asmagill commented:
Ok, I want to run some tests against the virtual touchbar and try seeing if we can detect in a nicer way whether or not `normalizedPosition` is available (without using `@try/@catch/@finally`), so I'll probably have another version I want you to test later tonight or tomorrow.  I'll also checkout the new links to see what they may have to offer, and it appears that there are some new event types that should probably be added as well.

But we have confirmation about `type` -- "direct" means touchbar, "indirect" means trackpad. Still wish I knew about `resting`, but I guess you can't have everything 😜

latenitefilms commented:
I'll try removing the `@try/@catch/@finally` and see if I can get it to crash on my machine.

Resting touches occur when a user simply rests their thumb on the trackpad device if that's what you're asking?

latenitefilms commented:
I'm also seeing that error on the Touch Bar:

```
2020-09-24 11:05:50: 11:05:50 ERROR:   LuaSkin: hs.eventtap callback error: NSInternalInconsistencyException: Cannot get normalizedPosition for this type of NSTouch
(
	0   CoreFoundation                      0x00007fff4b9ebacd __exceptionPreprocess + 256
	1   libobjc.A.dylib                     0x00007fff760f1a17 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff4ba067e6 +[NSException raise:format:arguments:] + 98
	3   Foundation                          0x00007fff4dca7791 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 194
	4   AppKit                              0x00007fff49320e12 -[NSTouch normalizedPosition] + 183
	5   event.so                            0x000000011161cccd NSTouch_toLua + 2173
	6   LuaSkin                             0x0000000101a2631b -[LuaSkin pushNSObject:withOptions:alreadySeenObjects:] + 2859
	7   LuaSkin                             0x0000000101a2d270 -[LuaSkin pushNSSet:withOptions:alreadySeenObjects:] + 2624
	8   LuaSkin                             0x0000000101a276b9 -[LuaSkin pushNSObject:withOptions:alreadySeenObjects:] + 7881
	9   LuaSkin                             0x0000000101a15193 -[LuaSkin pushNSObject:withOptions:] + 883
	10  LuaSkin                             0x0000000101a14d9a -[LuaSkin pushNSObject:] + 442
	11  event.so                            0x000000011162544c eventtap_event_getTouches + 1548
	12  LuaSkin                             0x0000000101b8c3cc luaD_call + 5932
	13  LuaSkin                             0x0000000101b11094 luaV_execute + 356420
	14  LuaSkin                             0x0000000101b8e05b luaD_call + 13243
	15  LuaSkin                             0x0000000101b8e880 luaD_callnoyield + 512
	16  LuaSkin                             0x0000000101bd2c0c f_call + 428
	17  LuaSkin                             0x0000000101a536ed luai_objcttry + 445
	18  LuaSkin                             0x0000000101b7fd42 luaD_rawrunprotected + 1922
	19  LuaSkin                             0x0000000101b94854 luaD_pcall + 836
	20  LuaSkin                             0x0000000101bd13e0 lua_pcallk + 5424
	21  LuaSkin                             0x0000000101a0b87b -[LuaSkin protectedCallAndTraceback:nresults:] + 827
	22  internal.so                         0x000000010f648556 eventtap_callback + 2582
	23  SkyLight                            0x00007fff70bee2a2 _ZL19processEventTapDataPvjjjPhj + 631
	24  SkyLight                            0x00007fff70aee290 _XPostEventTapData + 277
	25  SkyLight                            0x00007fff70bedfcd _ZL22eventTapMessageHandlerP12__CFMachPortPvlS1_ + 132
	26  CoreFoundation                      0x00007fff4b94d449 __CFMachPortPerform + 246
	27  CoreFoundation                      0x00007fff4b94d347 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE1_PERFORM_FUNCTION__ + 41
	28  CoreFoundation                      0x00007fff4b94d2a5 __CFRunLoopDoSource1 + 527
	29  CoreFoundation                      0x00007fff4b93529c __CFRunLoopRun + 2524
	30  CoreFoundation                      0x00007fff4b93466e CFRunLoopRunSpecific + 455
	31  HIToolbox                           0x00007fff4ab931ab RunCurrentEventLoopInMode + 292
	32  HIToolbox                           0x00007fff4ab92ded ReceiveNextEventCommon + 355
	33  HIToolbox                           0x00007fff4ab92c76 _BlockUntilNextEventMatchingListInModeWithFilter + 64
	34  AppKit                              0x00007fff48f2a77d _DPSNextEvent + 1135
	35  AppKit                              0x00007fff48f2946b -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1361
	36  AppKit                              0x00007fff48f23588 -[NSApplication run] + 699
	37  AppKit                              0x00007fff48f12ac8 NSApplicationMain + 777
	38  Hammerspoon                         0x00000001000116f4 main + 52
	39  libdyld.dylib                       0x00007fff778c03d5 start + 1
	40  ???                                 0x0000000000000005 0x0 + 5
)
```

However, this seems to work around it:

```objc
    if (![touch respondsToSelector:@selector(type)] || (NSInteger) [touch performSelector:@selector(type)] == 1 /* NSTouchTypeIndirect */) {
        NSPoint np = touch.normalizedPosition ;
        [skin pushNSPoint:np] ;
        lua_setfield(L, -2, "normalizedPosition") ;
    }
```

latenitefilms commented:
It also looks like there's an undocumented `force` value?

https://github.com/onmyway133/Runtime-Headers/blob/master/macOS/10.12/AppKit.framework/NSTouch.h

latenitefilms commented:
NSTouchTypeDirect = A direct touch from a user's finger on a screen.
NSTouchTypeIndirect = An indirect touch that is not on a screen, like a digitizer touch.

asmagill commented:
My comment re `resting` is that I've never seen it be any value but false. I would assume it means what you think (and the docs suggest so), but on my laptop (pre force-touch) and on my external force touch trackpad I never see it as anything other than `false`, as well as the snippits you've posted here (though if you *have* seen it as `true`, let me know under what conditions so we can confirm). Knowing for which devices it *can* return true would enhance the documentation.

As a general rule, I like to try to observe or force all combinations (at least as many as reasonably possible) so the docs can be made clear and we don't get questions like "why doesn't this value change when I do X?"

latenitefilms commented:
On my phone, so haven't checked your code yet, but as a thought have you enabled resting?

`[self setWantsRestingTouches:YES]`

latenitefilms commented:
This thread might spark some ideas:

https://stackoverflow.com/a/3792002

asmagill commented:
Ok, good enough for me to document that it will only be useful with certain apps that have enabled rest detection.

Going to try to return to this tomorrow (I needed some down time!) so may or may not make it into 81, depending upon when @cmsj wants to land that. I'm ok with this being after as there are enough fixes in 81 to merit it as the (hopefully) final fix to 79.

latenitefilms commented:
I'll have a play with combining this branch with the gesture simulation branch, and see if it can help with the documentation.

asmagill commented:
Your call.

Just so you know, I was planning to do at least the following (in no particular order) before considering this ready, though I suppose these could just as easily be future pulls:

* verify whether it works/doesn't work with the virtual touchbar. (since this is not a core module, it's not a high priority)
* check out the relevant headers to see what new event types need to be added to the `hs.eventtap.event.types` table.... again, not the highest priority because if you know the integer equivalent, you can specify that (I know to check `CGEventTypes.h` and `NSEvent.h`, but not sure if they are any others that should also be checked yet)
* per your observation above with trying without it, change the `@catch/@try/@finally` to something along the lines of `if (touch.type == NSTouchTypeIndirect) then { [skin pushNSPoint:touch.normalizedPosition] ; lua_setfield(L, -2, "normalizedPosition") ; end` and update docs saying that this property only exists for trackpads
* experiment with finding a way to see if other events *might* also contain `NSTouch` objects... the docs say only NSEventTypeGesture does, but as we see NSEventTypeDirectTouch does as well... is there a way to test without wrapping `asNSEvent.allTouches` in a `@try/@catch/@finally` block (because I already know it will throw an exception for `kCGEventMouseMoved` events)? Don't know yet...
* check out the links you give and see what they might suggest or inspire
* check for undocumented properties (like `force` you mention above) and see what we find

None of these prevent this from being at least somewhat useful already, so if you have your additions almost ready, merging what is here (though do fix the `normalizedPosition` one if you do so it doesn't use `@try/@catch/@finally`) makes a certain amount of sense. I can do the rest later.

latenitefilms commented:
Sorry - when I said "combining this branch", I just meant for the purposes of testing - I wasn't planning on doing a permanent merge. I think it's best to keep these pull requests seperate. In theory though, I can use the work you've done here, to observe the simulated gestures - which might give me a better understanding as to what they're actually doing.

asmagill commented:
Ahh... your documentation questions...

Ok, I plan to get back to this tomorrow, so who knows, they may both be ready around the same time. There is *some* overlap in this with `hs._asm.touchdevice` (this way at getting at the touch data is cleaner, but not quite as detailed, though I don't know how much of the additional information is actually that useful to us) so I may look at it for inspiration as well.

asmagill commented:
*Note to self, use Hopper to look for NSEvent (in AppKit, I believe) and see if a list of "safe" events for `allTouches` can be determined from there... how does it decide when to throw exception?*

latenitefilms commented:
@asmagill - Testing this further, this line:

```objc
if (asNSEvent.type == NSEventTypeGesture || asNSEvent.type == NSEventTypeDirectTouch) {
```

...should actually be something along the lines of:

```objc
if (asNSEvent.type == NSEventTypeGesture || asNSEvent.type == NSEventTypeDirectTouch || asNSEvent.type == NSEventTypeMagnify || asNSEvent.type == NSEventTypeSwipe || asNSEvent.type == NSEventTypeRotate || asNSEvent.type == NSEventTypeBeginGesture || asNSEvent.type == NSEventTypeEndGesture) {
```

...in order to pickup all the various touch events.

latenitefilms commented:
I'm also seeing `asNSEvent.type == 36`, but I have no idea what this is, as it's not defined in [`NSEvent.h`](https://github.com/phracker/MacOSX-SDKs/blob/master/MacOSX10.15.sdk/System/Library/Frameworks/AppKit.framework/Versions/C/Headers/NSEvent.h)?

asmagill commented:
If I may ask, how did you come up with the list of events? I suspected these might also work, but hadn't settled on a specific methodology to test them yet. Just curious...

latenitefilms commented:
I just added a `NSLog()` after the `else` to see what `NSEvents` were being skipped over as I messed around with the Track Pad and Touch Bar. I think the 36 was being triggered somehow by the "simulated gestures".

asmagill commented:
* brought this up to date with master, so posting to application fix is included now (no merge conflicts anymore)
* added some events to the `allTouches` whitelist that digging through the macOS Header files suggest *should* be touch safe. Haven't tested them yet, though, so let me know if you get any crashes with any of them:
    * NSEventTypeGesture
    * NSEventTypeMagnify
    * NSEventTypeSwipe
    * NSEventTypeRotate
    * NSEventTypeBeginGesture
    * NSEventTypeEndGesture
    * NSEventTypeSmartMagnify
    * NSEventTypeQuickLook
    * NSEventTypePressure
    * NSEventTypeDirectTouch
* added undocumented optional boolean to `allTouches` so you can experiment with other events -- it will bypass the whitelist if you specify an explicit `true` (I still can't identify 36, so @latenitefilms if you want to watch for and test this, you won't need to modify the source to do so -- just use `:allTouches(true)`.) Not sure if I'll leave this in when this is finally merged, but for now it saves recompile time.
* added some fields to the NSTouch array:
    * force - I suspect this may bomb on the touchbar -- @latenitefilms please try it out and let me know
    * previousNormalizedPosition
    * timestamp
* added some events and properties that we were missing
* began a long overdue reorganization of the properties and types tables

And I plan to review the `NSEvent` documentation further... I suspect there are some methods in there we can leverage that will make getting some details for certain events much easier than the current requirement of using `getProperty`. The event submodule of eventtap is really long overdue for an overhaul...

And a question:  Most of the newer (i.e. NSEvent inspired) additions of event types are listed as their macOS name in the `hs.eventtap.event.types` table. Originally I think I did this because I considered them a very experimental addition to the CoreGraphics approach everything else in the module took and wasn't even certain that they would be permanent additions. But compared to our standard camelCasing with "programmer branding" removed style used for everything else in Hammerspoon, they're kinda ugly. If I changed them into the style we use everywhere else in Hammerspoon (e.g. `NSEventTypeDirectTouch` would be `directTouch`) it would be a breaking change for people who use the table as the source when defining eventtaps. Should we do so? I'm leaning towards yes, but want to know what you guys think (@cmsj, if you haven't been following, I'd like your input, at least on this point, as well).

latenitefilms commented:
Here's some tests with the Track Pad:

https://gist.github.com/latenitefilms/3ee869d06472d050e8acbc306f3802cb

Here's some tests with the Touch Bar:

https://gist.github.com/latenitefilms/1cbea3bc5d5cbe2f41dfa6b8ce37c922

I did see Xcode crash with the Touch Bar if I try and do a "magnify" movement with two fingers - however interestingly, in the gist above, it didn't actually crash it just wrote those errors to the Console, so not sure why sometimes it causes a crash, and sometimes it writes to Console?

```
exception	NSException *	name: @"NSInternalInconsistencyException" - reason: @"622807001.765807 Missing Touches. This is a bug in IOHIDFamily"	0x0000604001315a10
```

FWIW - I prefer the camel case method.

cmsj commented:
@asmagill ultimately I'm fine with either option, but I think I'd lean more towards consistency within a module, than between modules.

asmagill commented:
@cmsj, just to be clear, the CoreGraphics defined event types are in our traditional camelCase; the NSEvent inspired ones currently are not. From the users perspective, there is no technical difference (they can be mixed when creating a new eventtap), and are even stored in the same reference table (`hs.eventtap.event.types`). 

Since the CG events were the original ones (copied over from the Mjolnir days) I'm assuming you're in favor of maintaining that consistency?

asmagill commented:
@latenitefilms latest update adds `location` and `previousLocation` to touches when `type` == "direct". Can't quite tell what the scale is...

`displaySize` for touchpads is 72PPI, so you multiply the normalized position by 72 to get the location in inches (if you want it, that is)

`displaySize` for the virtual touchbar seems to be the touchbar size (e.g. `{ __luaSkinType = "NSSize", h = 30.0, w = 1085.0 }`, but I'm getting an `x` value for location that seems to go from 0 to about 320 or so... can you check with a real touchbar and see what type of relationship, if any, you can discern? (Oh and `y` value of location seems to always be 15... guess that doesn't matter as much for a "bar", but the `x` might be useful.)

latenitefilms commented:
Given...
 
```
'NSEventTypeChangeMode' is only available on macOS 10.15 or newer
```

It might be worth using an `@available` check for:

```objc
lua_pushinteger(L, NSEventTypeChangeMode);         lua_setfield(L, -2, "NSEventTypeChangeMode");
```

latenitefilms commented:
Here's some more tests using the Touch Bar:

https://gist.github.com/latenitefilms/4bcef5b7d10d41e1fd8f03ce59e0fc50

asmagill commented:
Yeah, I haven't gotten around to dealing with the myriad of warnings eventtap throws up when I compile it outside of Hammerspoon for quicker turn-around time... it's on my list 😁

I also came across something about some events (including that one btw) that can't be watched for directly, but will only show up when you set the eventtap to watch for all events and then filter out the ones you want/don't want within the callback... I want to confirm this and see about updating `hs.eventtap.new` if so... watching for *everything* and having lua make the decision about what to respond to can cause things to slow down a lot so I'd rather shove such filtering into the Objective-C side of things, if I can isolate the specific events... I think it has to do with the Core Graphics event mechanism inheriting a lot of baggage from the 32bit world, so hopefully there is a clear demarcation and it only affects events with a type greater than 31, but I haven't corroborated this yet.

asmagill commented:
Can you test the touchbar again making sure to touch it as far to the left and right as you can and still get events triggered? You're definitely getting larger numbers than I am, but I've moved to using the touchbar generated by Duet on my iPad (it generates gesture events, while I can only seem to generate mouse events reliably from our virtual one... add that to the things I'd like to figure out but don't have time to look into for a while). But Duet may be doing its own scaling, so your physical touchbar is more reliable for determining and documenting how to scale the location values against the deviceSize in general.

latenitefilms commented:
Here you go, far left and far right:

https://gist.github.com/latenitefilms/0df0cfc31b56d8d5ad3287941de15608

latenitefilms commented:
I'm using a MacBook Pro (15-inch, 2017).

asmagill commented:
Interesting... you seem to see locations topping out just under 320 as well...

asmagill commented:
Do we know if the touchbar is considered a retina display, and if so what the scale factor is? I'm still trying to figure out the relationship between `location.x` and `deviceSize`  but haven't come up with a good explanation yet...

latenitefilms commented:
Not sure if this helps, but if I take a screenshot of my Touch Bar (CMD+SHIFT+6), it returns a 2170x60 @ 144dpi image?

<img width="1085" alt="Touch Bar Shot 2020-09-28 at 10 23 48 am" src="https://user-images.githubusercontent.com/22286696/94379609-df5a4000-0174-11eb-9a79-ee2290282900.png">

latenitefilms commented:
Thinking about it, it may be related to the main screen size?

If I take a screenshot of my laptop screen it's currently 3360x2100 @ 144dpi.

For example, maybe the reason the numbers start higher when you touch the far left of the Touch Bar is because the Touch Bar is centre of the screen? No sure - just food for thought.

asmagill commented:
Ok comparing the image to the reported `deviceSize` suggests that the touchbar has a scaling factor of 2, which I think is standard for Retina devices (but not the 4+k ones, IIRC).

Not sure if that helps or not, though...

What I'm trying to figure out is how to document it... with the trackpad and `normalizedPosition`, you can figure out where on the trackpad the touch is by doing something like:

~~~lua
    -- docs say trackpad is a 72 PPI device and that tracks with measuring my laptop's and external
    -- forcetouch trackpads with a ruler
    touchPosInInches = {
        x = touch.deviceSize.w * touch.normalizedPostion.x / 72, 
        y = touch.deviceSize.h * touch.normalizedPostion.y / 72,
    }
~~~

But I'm not sure how to do something similar with the touchbar... if we even can. The `deviceSize` says the width is 1085, your picture says it's 2170 (1085 * 2), so that tracks. Your earlier gist showed the largest `location.x` as 317, IIRC, so it's reasonable to think it's around 320ish, but I can't find a relationship between that max and the device width that makes sense... unless *maybe* we're not getting the location from the correct view of the touchbar... but using `nil` as the view *is* what I've found in online examples.

Ok, well, at least your reported location numbers track with mine using Duet, so I can dig deeper on my end with some assurance that what I find is likely to work for the physical touchbar as well.

latenitefilms commented:
Does this help at all?

https://lists.apple.com/archives/cocoa-dev/2017/Apr/msg00053.html

Specifically:

```objc
// Normalized position. From Apple's macOS Sierra 10.12.2 AppKit Release Notes: "Direct touches do not have a normalizedPosition in the digitizer space. Instead their position is in view coordinates.... AppKit will throw an exception if you attempt to acquire the wrong type of position for the type of touch."
        // NSPoint touchPoint = [thisTouch normalizedPosition]; // this will crash if uncommented
        // NSLog(@"normalized position: %@", NSStringFromPoint(touchPoint));

        // Location. From the NSTouch.h header file: "These two methods are only valid for Direct touches. A nil view means the touch location in the root container of touch." Experimentation shows that the "root container of Touch" means the Touch Bar device. The y-coordinate is not meaningful; use the x-coordinate to locate the distance of the touch (at 72 dpi) from the left end of the Touch Bar. The width of the deviceSize value is the maximum horizontal size of the Touch Bar.
        NSPoint touchLocation = [thisTouch locationInView:nil];
        NSPoint touchPreviousLocation = [thisTouch previousLocationInView:nil];
        NSLog(@"location: %@; previous location: %@", NSStringFromPoint(touchLocation), NSStringFromPoint(touchPreviousLocation));
```

latenitefilms commented:
Maybe also related:

https://github.com/erikolsson/Touch-Bar-Lemmings/blob/master/TouchLemmings/LemmingsScene.swift

asmagill commented:
I think that's one of the places where I saw to use `nil` as the specified view for the location... but if the max `x` is 320, and your touchbar is 1085 points wide at 72 dpi (remember it was 2170 at 144 dpi according to the image) then... how do they relate? We're missing something or something's changed since that post was originally made that isn't being taken into account.

latenitefilms commented:
Ummm, good question.

I just tried running the tests again, and I'm seeing same results:

https://gist.github.com/latenitefilms/a8684ea169ba22a823c8074ad4431b47

I'll ask around, and do some more Googling of existing Touch Bar apps to see what I can uncover.

latenitefilms commented:
FYI - I've reached out to Bill Cheeseman, and he's happy to offer guidance - but won't be for a little while as he's flat out gearing up for the Big Sur & Apple Silicon releases. I've also reached out to Andreas Hegenberg @ BetterTouchTool but haven't heard back yet.

asmagill commented:
Ok, gesture detection has been fleshed out. It appears that *specific* gestures can't be identified by CoreGraphics event handling, but are all subsumed under the single type of "gesture" (NSEventTypeGesture), however if you check the type of the NSEvent created from each captured CGEventRef, you get a more specific gesture type... this is now documented and `hs.eventtap.event:getType` now allows for an optional argument that tells it to return the specific NSEvent type.

I'm not seeing a discrepancy between direct touch location and the size of my touchbar, so maybe it was an issue with Catalina? At any rate, that seems to be working correctly now, and I added a helper function to get some gesture event details without having to pull in the entire array of touches.

I've not seen a `swipe` gesture event, though, so that's still a question mark, but then in testing @latenitefilms #2512 synthetic swipes aren't working for me either, so... if anyone *does* see them, I'd like to compare notes and see what is different.

asmagill commented:
@cmsj, this is ready to land. The outstanding swipe concern can't be tested unless/until we determine if we *can* generate synthetic swipes, but otherwise the rest of the changes work fine.
